[MachineSSAUpdater][AMDGPU] Add faster version of MachineSSAUpdater class.

vpykhtin opened the pull request:
This is a port of SSAUpdaterBulk to machine IR minus "bulk" part. Phi deduplication and simplification are not yet implemented but can be added if needed.

When used in AMDGPU to replace MachineSSAUpdater for i1 copy lowering, it reduced compilation time from 417 to 180 seconds   for the pass on a large test case (56% improvement).

cdevadas commented on llvm/include/llvm/CodeGen/MachineSSAUpdater2.h:
```diff
@@ -0,0 +1,78 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the MachineSSAUpdater2 class.
```
The filename MachineSSAUpdater2 doesn't look good. Can you give a meaningful suffix?

arsenm commented on llvm/include/llvm/CodeGen/MachineSSAUpdater2.h:
```diff
@@ -0,0 +1,78 @@
+
+  /// Indicate that a rewritten value is available in the specified block
+  /// with the specified value. Must be called before invoking Calculate().
+  void AddAvailableValue(MachineBasicBlock *BB, Register V) {
```
Start with lowercase

arsenm reviewed (commented):
Can you templatify the IR one like is done for other passes, e.g uniformity info?

vpykhtin commented:
> We also use SSAUpdater in AMDGPUGlobalISelDivergenceLowering.cpp, test changes look fine, instructions are added at different time as far as I can see. Why not replace existing one with fast version?

Sure, but let's do it gradually.

vpykhtin commented:
> Can you templatify the IR one like is done for other passes, e.g uniformity info?

I'm not sure if it complicate things more than does good, but some places possibly can be templatified like 'computeLiveInBlocks'.

arsenm commented on llvm/lib/CodeGen/MachineSSAUpdater2.cpp:
```diff
@@ -0,0 +1,180 @@
+            .getReg(0);
+  }
+
+  for (auto *BBInfo : DomPath)
```
No auto, and braces

arsenm commented on llvm/lib/CodeGen/MachineSSAUpdater2.cpp:
```diff
@@ -0,0 +1,180 @@
+  SmallPtrSet<MachineBasicBlock *, 2> UsingBlocks(UseBlocks.begin(),
+                                                  UseBlocks.end());
+  SmallVector<MachineBasicBlock *, 32> IDFBlocks;
+  SmallPtrSet<MachineBasicBlock *, 32> LiveInBlocks;
```
This is probably too big for a SmalLPtrSet

vpykhtin replied:
Made 4 as default size.

vpykhtin commented on llvm/include/llvm/CodeGen/MachineSSAUpdater2.h:
```diff
@@ -0,0 +1,78 @@
+//===- MachineSSAUpdaterBulk.h - Unstructured SSA Update Tool ----------*- C++
```
Renamed to MachineIDFSSAUpdater.

vpykhtin commented on llvm/lib/CodeGen/MachineIDFSSAUpdater.cpp:
```diff
@@ -0,0 +1,181 @@
+  }
+
+  if (!V) {
+    V = createInst(TargetOpcode::IMPLICIT_DEF, TopDomBB,
```
Added parameter to the constructor to select appropriate opcode. Thought about doing this CRTP but it seems an overkill.

vpykhtin replied:
gentle ping.

arsenm commented on llvm/include/llvm/CodeGen/MachineIDFSSAUpdater.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef LLVM_TRANSFORMS_UTILS_MACHINE_SSAUPDATER2_H
+#define LLVM_TRANSFORMS_UTILS_MACHINE_SSAUPDATER2_H
+
+#include "llvm/CodeGen/MachineRegisterInfo.h"
```
Only need to forward declare MachineRegisterInfo

vpykhtin committed:
[MachineSSAUpdater][AMDGPU] Add faster version of MachineSSAUpdater class.

This is a port of SSAUpdaterBulk to machine IR minus "bulk" part. Phi
deduplication and simplification are not yet implemented but can be added if needed.

When used in AMDGPU to replace MachineSSAUpdater for i1 copy lowering, it
reduced compilation time from 417 to 180 seconds for the pass on a large test
case (56% improvement).

vpykhtin committed:
Merge branch 'main' into machine_ssa_updater2

vpykhtin committed:
Merge branch 'main' into machine_ssa_updater2

vpykhtin commented:
The failure doesn't seem related to this patch
